Doc/library/zoneinfo.rst

```diff
@@ -238,13 +238,6 @@ The following class methods are also available:
 
     .. TODO: Add "See `cache_behavior`_" reference when that section is ready.
 
-    .. warning::
```

Comment:
I don't think we can remove this. Maybe we need to change "process-wide" to something else, but if anything the fact that there are per-interpreter caches has made a warning about this function *more* important.

The problem this warning is describing comes from the fact that the weak `ZoneInfo` cache guarantees that `ZoneInfo(some_str)` always produces the *same object*, so that you can do stuff like this:

```python
datetime(2020, 12, 1, tzinfo=ZoneInfo("America/New_York")) - datetime(2020, 9, 1, tzinfo=ZoneInfo("America/New_York")
```

And have it be a [same-zone subtraction](https://blog.ganssle.io/articles/2018/02/aware-datetime-arithmetic.html).

The danger with `clear_cache` is that you might do something like this:

```python
def f(dt: datetime, offset: timedelta) -> Iterable[timedelta]:
    while True:
        yield dt - datetime(2020, 12, 1, tzinfo=zoneinfo.ZoneInfo("America/New_York"))
        dt += offset

offsets = f(datetime(2021, 6, 1, tzinfo=zoneinfo.ZoneInfo("America/New_York")), offset=timedelta(minutes=1))

print(next(offsets))
print(next(offsets))
zoneinfo.ZoneInfo.clear_cache()
print(next(offsets))
```

The result is this:

```
182 days, 0:00:00
182 days, 0:01:00
181 days, 23:02:00
```

In this case, `dt` still holds a reference to the original New York time zone, while the reference point in the function constructs a new one after the cache is cleared.

In this case it's a bit stark, but you can imagine in a complex operation you may have many `ZoneInfo` objects attached to various `datetime` objects, stored in lists, stored in dicts, etc, and by calling `ZoneInfo.clear_cache` you start accumulating a bunch of subtly incompatible `ZoneInfo` objects as new `ZoneInfo`s are created for the same zone.

Having a per-interpreter `zoneinfo` module doesn't change this, if anything it makes the accounting harder (since you have to call `ZoneInfo.clear_cache()` on *every* `ZoneInfo` class if you want it to be across the board).

In fact, I'm assuming here that it's not possible for objects from one `zoneinfo` instance to ever interact with objects from another `zoneinfo` instance in the multiple interpreter case, correct? You'd have to basically pickle them and send them between interpreters or something, right? If not, then I fear the guarantees about the cache behavior might require us to do some more radical restructuring of the way the caches work.

Reply:
> In fact, I'm assuming here that it's not possible for objects from one zoneinfo instance to ever interact with objects from another zoneinfo instance in the multiple interpreter case, correct?

Yes, that's the point of [interpreter] isolation.

> You'd have to basically pickle them and send them between interpreters or something, right?

Yes, like before, you'd have to use a footgun; nothing changes here.

> If not, then I fear the guarantees about the cache behavior might require us to do some more radical restructuring of the way the caches work.

I don't see how per-interpreter caches changes anything. You can of course restructure how caches work, but that is definitely out of scope for this issue/PR :)

Reply:
> > You'd have to basically pickle them and send them between interpreters or something, right?
> 
> Yes, like before, you'd have to use a footgun; nothing changes here.

Actually, pickling a `zoneinfo.ZoneInfo` object will work just fine, because the unpickling logic hits the cache, so I don't think there's even a footgun if you want to send objects between interpreters.



> I don't see how ~per-interpreter~ per-module instance caches changes anything. You can of course restructure how caches work, but that is definitely out of scope for this issue/PR :)

Yeah, I just didn't understand the subinterpreter mechanism enough to understand that there's no way to do something like store a constructed `zoneinfo.ZoneInfo` in memory shared between Interpreter 1 and Interpreter 2. If that were the case, the invariant that `ZoneInfo`s constructed from the same key are always the same object would be broken. But of course it should have been pretty obvious that this is not possible, since all kinds of other stuff would also break (like, for example, `isinstance(some_variable, float)` would presumably break when `some_variable` is a float constructed in a different subinterpreter).